### PR TITLE
Apply attr proc macros before cfg processing

### LINKED ISSUE #39336: macros: don't `cfg`-process items before applying attribute proc macros
cc #38356

### PR DESCRIPTION
**Pull request by tmnilsson**
Fixes #39336.
r? @jseyfried

### CODE REVIEW

**Comment by aidanhs** [comment]
Thanks for the PR @tmnilsson, we'll check in now and again to make sure @jseyfried gets to this soon!

**Comment by carols10cents** [comment]
review ping @jseyfried! pinging you on IRC too!

File: src/libsyntax/ext/expand.rs [hunk 1]
```diff
@@ -867,7 +867,12 @@ impl<'a, 'b> Folder for InvocationCollector<'a, 'b> {
 
         let (attr, traits, mut item) = self.classify_item(item);
         if attr.is_some() || !traits.is_empty() {
-            let item = Annotatable::Item(fully_configure!(self, item, noop_fold_item));
+            let item = if !attr.is_some() {
+                Annotatable::Item(fully_configure!(self, item, noop_fold_item))
+            }
+            else {
```
Comment (jseyfried) [inline src/libsyntax/ext/expand.rs]: nit: `} else {` is more idiomatic

File: src/libsyntax/ext/expand.rs [hunk 2]
```diff
@@ -867,7 +867,12 @@ impl<'a, 'b> Folder for InvocationCollector<'a, 'b> {
 
         let (attr, traits, mut item) = self.classify_item(item);
         if attr.is_some() || !traits.is_empty() {
-            let item = Annotatable::Item(fully_configure!(self, item, noop_fold_item));
+            let item = if !attr.is_some() {
```
Comment (jseyfried) [inline src/libsyntax/ext/expand.rs]: nit: `!attr.is_some()` -> `attr.is_none()`

**Comment by jseyfried** [comment]
@tmnilsson looks good so far!

Could you test that 
`#[derive(Foo)] #[foo] struct S { #[cfg(any())] x: i32 }` gets fully configured (i.e. `x` is removed), given
```rust
#[proc_macro_derive(Foo, attributes(foo))]
fn derive_foo(input: TokenStream) -> TokenStream { input }
```

The issue here is that when collecting `#[derive(Foo)] #[foo] struct S { ... }`, we don't know whether `#[foo]` is an attribute proc-macro or just an "inert" attribute used to give information to `#[derive(Foo)]`, so we collect is as an attribute proc macro first (i.e. your code won't fully configure).

If we learn that `#[foo]` is not in fact a proc macro, we re-collect into a derive invocation here: https://github.com/tmnilsson/rust/blob/835992d4b14800cac0c450f893c050fdedd18dca/src/libsyntax/ext/expand.rs#L285.

I believe you'll need to `fully_configure!` the `item` at this line to ensure that we don't regress examples like this that compile today.

**Comment by tmnilsson** [comment]
Thanks for the feedback! That example would not get fully configured, as you expected. I wasn't even aware of inert attributes, actually. Anyway, I moved the full configure as you suggested, and that seems to make it work.

File: src/libsyntax/ext/expand.rs [hunk 3]
```diff
@@ -400,6 +400,27 @@ impl<'a, 'b> MacroExpander<'a, 'b> {
...
+            features: self.cx.ecfg.features,
+        };
+        // Since the item itself has already been configured by the InvocationCollector,
+        // we know that fold result vector will contain exactly one element
+        match item {
+            Annotatable::Item(item) => {
+                Annotatable::Item(cfg.fold_item(item).pop().unwrap())
+            }
+            Annotatable::TraitItem(item) => {
+                Annotatable::TraitItem(P(cfg.fold_trait_item(item.unwrap()).pop().unwrap()))
```
Comment (jseyfried) [inline src/libsyntax/ext/expand.rs]: nit: `item.map(|item| cfg.fold_trait_item(item).pop().unwrap())` is more idiomatic and efficient here (i.e. with `P`). Same for `ImplItem`.

**Comment by jseyfried** [comment]
@tmnilsson Excellent, thanks!
r=me modulo optional nit.

File: src/libsyntax/ext/expand.rs (hunk 3 above)
Comment (tmnilsson) [reply src/libsyntax/ext/expand.rs]: Got it, fixed.

**Pull request merged**